Classes/Message/MessageMetadata.php

```diff
@@ -59,25 +58,28 @@ public function getProperties(): array
      * @param string $name
      * @param mixed $value
      * @return MessageMetadata
-     * @throws MessageMetadataException
      */
-    public function add(string $name, $value)
+    public function withProperty(string $name, $value)
     {
-        if ($this->contains($value)) {
-            throw new MessageMetadataException(sprintf('The given value "%s" exist'), 1472853526);
-        }
-        $this->properties[$name] = $value;
-        return $this;
+        return new static(array_merge($this->properties, [$name => $value]), $this->timestamp);
     }
 
     /**
-     * @param string $name
+     * @param array $properties
+     * @return MessageMetadata
+     */
+    public function withProperties(array $properties)
+    {
+        return new static($properties, $this->timestamp);
+    }
+
+    /**
+     * @param array $properties
      * @return MessageMetadata
      */
-    public function remove(string $name)
+    public function andProperties(array $properties)
```

Comment:
… and while we're at it: let's follow @dfeyer's example and start using return type declarations (in this case for `MessageData`).